src/env.rs

```diff
@@ -0,0 +1,112 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+use std::sync::Arc;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::thread::{Builder, JoinHandle};
+
+use grpc_sys;
+
+use async::Promise;
+use cq::{CompletionQueue, EventType};
+
+// event loop
+fn poll_queue(cq: Arc<CompletionQueue>) {
+    loop {
+        let e = cq.next();
+        match e.event_type {
+            EventType::QueueShutdown => break,
+            // timeout should not happen in theory.
+            EventType::QueueTimeout => continue,
+            EventType::OpComplete => {}
+        }
+
+        let ctx: Box<Promise> = unsafe { Box::from_raw(e.tag as _) };
+
+        ctx.resolve(&cq, e.success != 0);
+    }
+}
+
+/// An object that used to control concurrency and start event loop.
+pub struct Environment {
+    cqs: Vec<Arc<CompletionQueue>>,
+    idx: AtomicUsize,
+    _handles: Vec<JoinHandle<()>>,
+}
+
+impl Environment {
+    /// Initialize grpc and create a threadpool to poll event loop.
+    ///
+    /// Each thread in threadpool will has one event loop.
+    pub fn new(cq_count: usize) -> Environment {
+        assert!(cq_count > 0);
+        unsafe {
+            grpc_sys::grpc_init();
+        }
+        let mut cqs = Vec::with_capacity(cq_count);
+        let mut handles = Vec::with_capacity(cq_count);
+        for i in 0..cq_count {
+            let cq = Arc::new(CompletionQueue::new());
+            let cq_ = cq.clone();
+            let handle = Builder::new()
+                .name(format!("grpcpollthread-{}", i))
+                .spawn(move || poll_queue(cq_))
+                .unwrap();
+            cqs.push(cq);
+            handles.push(handle);
+        }
+
+        Environment {
+            cqs: cqs,
+            idx: AtomicUsize::new(0),
+            _handles: handles,
+        }
+    }
+
+    /// Get all the created completion queues.
+    pub fn completion_queues(&self) -> &[Arc<CompletionQueue>] {
+        self.cqs.as_slice()
+    }
+
+    /// Pick an arbitrary completion queue.
+    pub fn pick_a_cq(&self) -> Arc<CompletionQueue> {
```

Comment:
I would suggest the removal of "a" since it sounds odd as a function name. But this is more a matter of convention, if you have "a" in other similar situations, keep it so that it could be consistent.